### PR TITLE
cleanup_before: untap unneeded taps on GitHub-hosted runners

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
Should help us get rid of the strange warnings seen in [^1]. `aws/tap/aws-sam-cli` calls `opoo` unconditionally once evaluated [^2]. The tap `aws/tap` is installed on GitHub-hosted runners, and we evaluate all formulae when determining the dependents of testing formulae.

A more ideal and comprehensive solution would be to untap all unneeded taps by (recursively) determining which taps are actually needed by the formulae in the testing tap, but that requires evaluating all formulae in the testing tap and can be a bit more work. And in the end, for third-party taps, doing that will still not eliminate the warnings if a dependency has an unconditional `opoo` call like that. So for now, I'm restricting the cleanup to `homebrew/core`.

[^1]: https://github.com/Homebrew/homebrew-core/actions/runs/11148695947/job/30986582554?pr=192640
[^2]: https://github.com/aws/homebrew-tap/blob/f2dcc08fa89256b3893a6cfc82e3ac71b05ce28e/Formula/_aws-sam-cli.rb#L58

### CODE REVIEW

**Review by p-linnane — APPROVED** [review]
Awesome, thanks!

File: lib/tests/cleanup_before.rb [hunk 1]
```diff
@@ -48,6 +50,11 @@ def run!(args:)
         cleanup_shared
 
         installed_taps = Tap.select(&:installed?).map(&:name)
+        if on_github_hosted_runner && tap.to_s == CoreTap.instance.name
+          (installed_taps - REQUIRED_TAPS).each do |tap|
+            test "brew", "untap", tap
+          end
+        end
```
Comment (Bo98) [inline lib/tests/cleanup_before.rb:57]: Won't this untap homebrew/core?

Comment (ZhongRuoyu) [reply lib/tests/cleanup_before.rb:57]: Whoops -- I misread `REQUIRED_TAPS` as there was a `CoreTap.instance.name` below.

**Comment by MikeMcQuaid** [comment]
Thanks @ZhongRuoyu, this needed reverted but I still think this approach is worth pursuing!

> is already meant to do this but I guess it isn't working?

Yes, I think it'd be good to fixup that logic to handle this as expected.

**Pull request merged**